### ISSUE TITLE
Potential bug in the misc.py file when creating masks for training

### ISSUE THREAD

**Issue by AceVenturos** [ChristophReich1996/Semantic_Pyramid_for_Image_Generation#3]
The paper states in section 3.2, that the default mode of training is to select 1 layer and mask out the others. I believe with this implementation, all of the layers after the selected layer will not be masked out as you use torch.ones instead of torch.zeros for the default cases.

I've uploaded a small dcoument that discusses this. There was also an edge case where if the 1st layer(365,) there is a possibility the second layer will be randomised and the third layer may not use the spatially varying mask.

[Potential Mask Issue in Starter Code.pdf](https://github.com/ChristophReich1996/Semantic_Pyramid_for_Image_Generation/files/6122881/Potential.Mask.Issue.in.Starter.Code.pdf)

Note, due to computational resource limits, I've reduced the image size to 64x64 and the classes down to 10 so there is some variation in the code. Thanks for this implementation by the way, it has been a great help to me!

**Comment by ChristophReich1996** [comment]
Hi, thanks for raising this issue and for the clear explanation. Let me check this in detail and I keep you updated. Fell free to raise a pull request if you like to.

Cheers
Christoph

**Comment by ChristophReich1996** [comment]
Hey, so a quick update. Your concerns about the mask generation function were reasoned. I added your suggested changes to the repository. Thanks again for indicating this issue and for the clear explanation. Moreover, I implemented conditionalized batch normalization, as in the original self-attention GAN, to incorporate the class labels. The current commit is not tested now. However, I plan to do some tests this week. I keep you updated about new results.

PS: If you face more bugs, let me know. Additionally, if you achieved some improved results, I would be curious to know about them :)

Cheers
Christoph

**Comment by AceVenturos** [comment]
Hi Christoph, thanks for taking the time to review the mask generation function, I'm currently trying to achieve some better results but no luck so far and more tuning is needed on my part as I'm working with a reduced class dataset compared to the original paper with 64x64 images due to my current resources. 

If I achieve any better results I'll update you, I would be interested likewise if you achieve some improved results, particularly with the addition of the conditionalized batch normalisation :)

Thanks, 
Jamie

**Comment by ChristophReich1996** [comment]
Hi, I quick update; I guess the bad results come from the resolution I used. The resolution I used was 256X256. However, the VGG-16 was pre-trained on 224X224. Thus, the higher layers' activations may not capture the semantics of the input image (especially the FC layers). Currently, I fine-tune the VGG-16 on the higher resolution. Hopefully, using this fine-tuned model yields better overall results.

Cheers,
Christoph

**Comment by AceVenturos** [comment]
Hey Christoph,

Hope you're doing well, I was trying to implement some of your nwq changes to try and improve the results. One issue I'm having is with kornia, particularly within the VGG16 model's forward function when the input is normalised:

```
output = kornia.normalize(output,
                mean=torch.tensor([0.485, 0.456, 0.406], device=output.device),
                std=torch.tensor([0.229, 0.224, 0.225], device=output.device)) 
```

Because of limited resource I've had to reduce the image size down to 64 so I have adapted your code to work with smaller images. The 'output' tensor is shape [3, 64, 64] before the normalise function, the function is throwing an error:

**Comment by AceVenturos** [comment]
Hey Christoph,

If you have a minute I'd really appreciate if you could answer this question, I see that you implemented conditional batch normalisation recently, before this was the generator/discriminator not being conditioned by the labels? I.e. was the model unsupervised?

If it's any better, I could switch over to email or an alternative to message you? My email is jmckeown35@qub.ac.uk

Thanks,
Jamie

**Comment by ChristophReich1996** [comment]
Hi Jamie,

so let's first clarify how supervised and unsupervised learning are defined. 

In the **supervised learning** approach, the goal is to learn a mapping from inputs <img src="https://render.githubusercontent.com/render/math?math=x"> to outputs <img src="https://render.githubusercontent.com/render/math?math=y">, given a labeled set of input-output pairs <img src="https://render.githubusercontent.com/render/math?math=D=\{(x_i ,y_i )\}^{N}_{i=1}">. Here <img src="https://render.githubusercontent.com/render/math?math=D"> is called the training set, and <img src="https://render.githubusercontent.com/render/math?math=N"> is the number of training examples. [1]

In the **unsupervised learning** approach we are only given inputs, <img src="https://render.githubusercontent.com/render/math?math=D = \{x_i\}^{N}_{i=1}"> , and the goal is to find “interesting patterns” in the data. [1]

In easy words, generative adversarial networks (GANs) are, as proposed by Goodfellow et al., unsupervised [2]. This is because an implicit generative model is desired to be learned based on an unlabeled dataset. 
The "Semantic Pyramid for Image Generation" paper, however, utilizes a different GAN architecture [5]. Namely the Self-Attention Generative Adversarial Network (SAGAN) architecture [3]. On a high level, the SAGAN is a combination of the cGAN with projection discriminator [4] and self-attention layers. This cGAN with projection discriminator introduced a novel way to improve the generative performance by utilizing a projection layer to incorporate class labels into the discriminator. Additionally, conditionalized batch normalization (CBN) was utilized to incorporate the class labels into the generator network. Since no supervised signal is used in the loss to train the SAGAN I would consider the SAGAN (with or without CBN) as **unsupervised**. 
In the case of the "Semantic Pyramid for Image Generation" training approach it gets a bit more complicated. Since not only an adversarial loss is utilized, but also a semantic reconstruction loss is utilized. The semantic reconstruction loss employs the real image as a target label. Thus, I would describe the "Semantic Pyramid for Image Generation" training approach as a union of **adversarial (unsupervised) and self-supervised [8] learning**. [2, 3, 4, 5]
But be aware there are also GANs which are (semi-)supervised see [6, 7].

**Disclaimer:** This is only my view of things and I would recommend writing an email to Assaf Shocher for more detailed questions about the training approach.

Cheers
Christoph

````bibtex
[1] @book{Murphy2012,
    title={Machine learning: a probabilistic perspective},
    author={Murphy, Kevin P},
    year={2012},
    publisher={MIT press}
}
````

````bibtex
[2] @inproceedings{Goodfellow2014,
    title = {Generative Adversarial Nets},
    author = {Goodfellow, Ian and Pouget-Abadie, Jean and Mirza, Mehdi and Xu, Bing and Warde-Farley, David and Ozair, Sherjil and Courville, Aaron and Bengio, Yoshua},
    booktitle = {Advances in Neural Information Processing Systems},
    volume = {27},
    year = {2014}
}
````

````bibtex
[3] @inproceedings{Zhang2019,
    title={Self-attention generative adversarial networks},
    author={Zhang, Han and Goodfellow, Ian and Metaxas, Dimitris and Odena, Augustus},
    booktitle={International conference on machine learning},
    pages={7354--7363},
    year={2019}
}
````

````bibtex
[4] @article{Miyato2018,
    title={cGANs with projection discriminator},
    author={Miyato, Takeru and Koyama, Masanori},
    journal={arXiv preprint arXiv:1802.05637},
    year={2018}
}
````

````bibtex
[5] @inproceedings{Shocher2020,
    title={Semantic pyramid for image generation},
    author={Shocher, Assaf and Gandelsman, Yossi and Mosseri, Inbar and Yarom, Michal and Irani, Michal and Freeman, William T and Dekel, Tali},
    booktitle={Proceedings of the IEEE/CVF Conference on Computer Vision and Pattern Recognition},
    pages={7457--7466},
    year={2020}
}
````

````bibtex
[6] @inproceedings{Odena2017,
    title={Conditional image synthesis with auxiliary classifier gans},
    author={Odena, Augustus and Olah, Christopher and Shlens, Jonathon},
    booktitle={International conference on machine learning},
    pages={2642--2651},
    year={2017},
    organization={PMLR}
}
````

````bibtex
[7] @article{Sricharan2017,
    title={Semi-supervised conditional gans},
    author={Sricharan, Kumar and Bala, Raja and Shreve, Matthew and Ding, Hui and Saketh, Kumar and Sun, Jin},
    journal={arXiv preprint arXiv:1708.05789},
    year={2017}
}
````

````bibtex
[8] @article{jing2020self,
    title={Self-supervised visual feature learning with deep neural networks: A survey},
    author={Jing, Longlong and Tian, Yingli},
    journal={IEEE Transactions on Pattern Analysis and Machine Intelligence},
    year={2020},
    publisher={IEEE}
}
````

**Comment by AceVenturos** [comment]
Hi Christoph,

Thanks for taking the time to write a detailed response, I really appreciate it! Your description has definitely cleared a few things up for me. 

Cheers,
Jamie
